### PR TITLE
[WIP] new mouse acceleration model

### PR DESCRIPTION
**Pull request by JohanJu**
With this PR the behavior of pressing the opposite direction key will make the mouse slow down instead of going full speed in the other direction. Any feedback on what you think of this model is highly appreciated.

### CODE REVIEW

**Comment by drashna** [comment]
Travis CI found a critical error here: 
```
In file included from tmk_core/common/mousekey.c:24:0:
tmk_core/common/mousekey.h:54:30: error: large integer implicitly truncated to unsigned type [-Werror=overflow]
 #define MOUSEKEY_TIME_TO_MAX 500
                              ^
tmk_core/common/mousekey.c:50:26: note: in expansion of macro 'MOUSEKEY_TIME_TO_MAX'
 uint8_t mk_time_to_max = MOUSEKEY_TIME_TO_MAX;
```

**Comment by shieldsd** [comment]
>  uint8_t mk_time_to_max = MOUSEKEY_TIME_TO_MAX;
> ```

This needs to be uint16_t to avoid truncation when the value is greater than 255

**Comment by JohanJu** [comment]
Thanks for the reminder about the truncation. I'm mostlyinterested in if other see this as an improvement or rather keep

**Comment by DanielGGordon** [comment]
I don't know about the code, but I love this idea! I use the mouse keys all the time and would love to have better control.

**Comment by drashna** [comment]
There is a merge conflict with this now.  Could you update it?

**Comment by drashna** [comment]
Any updates on this?

**Comment by JohanJu** [comment]
> Any updates on this?

It's ready for review

**Comment by drashna** [comment]
One question then, could you make sure that if you hit two directions at once, that it respects both?   

eg, if you hit up and right, that the cursor goes diagonal? 

From a quick look, I don't think this should be an issue, but I wand to make sure.

**Comment by JohanJu** [comment]
Yes it goes diagonaly.

File: tmk_core/common/mousekey.c [hunk 1]
```diff
@@ -288,13 +280,158 @@ void mousekey_off(uint8_t code) {
...
+}
+
+void mousekey_on(uint8_t code)
+{
+    if (code == KC_MS_UP) {
+        if (y_state)  y_state = 2;
+        else          y_state = -1;
+    }
+    else if (code == KC_MS_DOWN) {
+        if (y_state)  y_state = -2;
```
Comment (yanfali) [inline tmk_core/common/mousekey.c]: can we replace this magic numbers with defines?

File: tmk_core/common/mousekey.c [hunk 2]
```diff
@@ -288,13 +280,158 @@ void mousekey_off(uint8_t code) {
...
+uint8_t mk_delay = MOUSEKEY_DELAY/10;
+uint8_t mk_interval = MOUSEKEY_INTERVAL;
+uint8_t mk_max_speed = MOUSEKEY_MAX_SPEED;
+uint8_t mk_slow_speed = MOUSEKEY_MAX_SPEED / MOUSEKEY_SLOW_RATIO;
+uint8_t mk_time_to_max = MOUSEKEY_TIME_TO_MAX;
+uint8_t acc_per_int = ((uint16_t) MOUSEKEY_MAX_SPEED) * MOUSEKEY_INTERVAL / MOUSEKEY_TIME_TO_MAX;
+uint8_t mk_wheel_max_speed = MOUSEKEY_WHEEL_MAX_SPEED;
+uint8_t mk_wheel_time_to_max = MOUSEKEY_WHEEL_TIME_TO_MAX;
+
+// x positive right, y positive down
```
Comment (yanfali) [inline tmk_core/common/mousekey.c]: these are magic numbers. Please turn them into defines

Comment (JohanJu) [reply tmk_core/common/mousekey.c]: I'm just following the current code.
https://github.com/qmk/qmk_firmware/blob/3a7c65b42442971bcb3369cc9dc3615b6aad1383/tmk_core/common/mousekey.c#L53-L64
https://github.com/qmk/qmk_firmware/blob/3a7c65b42442971bcb3369cc9dc3615b6aad1383/tmk_core/common/mousekey.c#L126-L130

File: tmk_core/common/mousekey.c (hunk 1 above)
Comment (JohanJu) [reply tmk_core/common/mousekey.c]: Do you mean the 2 and -1, I can change that to defines.

File: tmk_core/common/mousekey.c (hunk 2 above)
Comment (yanfali) [reply tmk_core/common/mousekey.c]: I was mostly concerned with the directions. I like the change you made. Much clearer to follow.

**Comment by yanfali** [comment]
@drashna are new defines appropriate? Otherwise lgtm

**Comment by drashna** [comment]
There are some merge conflicts that have popped up.

Could you update this?

**Pull request closed without merging**